internal/vm/validation.go

```diff
@@ -0,0 +1,225 @@
+package vm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	v1 "github.com/hashicorp/consul-api-gateway/internal/api/v1"
+	"github.com/hashicorp/consul-api-gateway/internal/common"
+	"github.com/hashicorp/consul-api-gateway/internal/vault"
+	consulapi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+)
+
+// Validator is responsible for validating the more complex fields of Gateways and Routes
+type Validator struct {
+	client      *consulapi.Client
+	vaultClient vault.KVClient
+	logger      hclog.Logger
+}
+
+func NewValidator(logger hclog.Logger, vaultClient vault.KVClient, client *consulapi.Client) *Validator {
+	return &Validator{
+		client:      client,
+		vaultClient: vaultClient,
+		logger:      logger,
+	}
+}
+
+// ValidateGateway validates a Gateway definition
+func (v *Validator) ValidateGateway(ctx context.Context, gateway *v1.Gateway) error {
```

Comment:
It feels like some of this logic is duplicated from the k8s validation and it would be desirable to ensure they stay in sync - should substantial parts of this (and the corresponding k8s logic) be moved out to a common layer above the `vm` and `k8s` packages?